File: graph/traffic_network.py
"""https://www.spoj.com/problems/TRAFFICN/en/"""
from queue import PriorityQueue

INF = int(1e9)

class Node:
	def __init__(self, id, cost):
		self.id = id
		self.cost = cost
	def __lt__(self, other):
		return self.cost <= other.cost

def Dijsktra(s):
	pq = PriorityQueue()
	pq.put(Node(s,0))
	dist[s] = 0

	while(not pq.empty()):
		node = pq.get()
		u = node.id
		w = node.cost

		# if u == e:
		# 	break
		for v in graph[u]:
			if dist[v.id] > w + v.cost:
				dist[v.id] = w + v.cost
				pq.put(Node(v.id, dist[v.id]))
	return dist[e]
#To see file traffic_network.cpp
if __name__ == "__main__":

	tc = int(input())
	while tc:
		n,m,k,s,t = list(map(int, input().strip().split(' ')))
		graph = [[] for _ in range(n+1)]
		dist = [INF for _ in range(n+1)]
		for i in range(m):
			u,v,w = list(map(int, input().strip().split(' ')))
			graph[u].append(Node(v,w))
		for i in range(k):
			u,v,w = list(map(int, input().strip().split(' ')))
			graph[u].append(Node(v,w))
			graph[v].append(Node(u,w))
		print(Dijsktra(s))
		tc -=1

File: graph/traffic_network.cpp
#include <iostream>
#include <queue>
#include <vector>
#include <unordered_map>
#include <string>
using namespace std;
const int INF = 0x3f3f3f3f;



struct option
{
	bool operator() (const pair<int, int>&a, const pair<int, int>&b) const
	{
		return a.second > b. second;
	}
};

void Dijsktra(int s, int e, vector<int> &dist, vector<vector<pair<int, int> > > & graph)
{
	priority_queue<pair<int, int>, vector<pair<int, int> > , option > pq;
	pq.push(make_pair(s,0));
	dist[s] = 0;
	while(!pq.empty())
	{
		pair<int, int> top = pq.top();
		pq.pop();
		int u = top.first;
		int w = top.second;
		// if(w > dist[u])
		// 	continue;
		// if(u == e)
		// 	break;
		for(int i = 0; i < graph[u].size(); ++i)
		{
			pair<int, int> neighbor = graph[u][i];
			if(dist[neighbor.first] >  w + neighbor.second)
			{
				dist[neighbor.first] = w + neighbor.second;
				pq.push(pair<int, int>(neighbor.first, dist[neighbor.first]));
			}
		}
	}
	// return dist[e];

}
int main()
{
	int n,m,k,s,t, tc;
	int u, v, w, shortest;
	int res;
	vector<int> dist_s;
	vector<int> dist_t;
	vector<vector<pair<int, int> > >  graph_s;
	vector<vector<pair<int, int> > >  graph_t;

	cin>>tc;
	while(tc--)
	{
		cin>>n>>m>>k>>s>>t;
		graph_s = vector<vector<pair<int, int> > > (n+1, vector<pair<int, int> > ());
		graph_t = vector<vector<pair<int, int> > > (n+1, vector<pair<int, int> > ());

		dist_s = vector<int>(n+1, INF);
		dist_t = vector<int>(n+1, INF);
		//one way road
		for(int i = 0; i < m; i++)
		{
			cin>>u>>v>>w;
			graph_s[u].push_back(pair<int, int>(v, w));

			graph_t[v].push_back(pair<int, int>(u, w));

		}
		Dijsktra(s,t, dist_s, graph_s);
		Dijsktra(t,s, dist_t, graph_t);
		res = dist_s[t];
		//two way road
		for(int i = 0; i < k ; i++)
		{
			cin>>u>>v>>w;

			shortest = min(dist_s[u] + dist_t[v] + w, dist_s[v] + dist_t[u] + w);
			res = min(shortest, res);

		}

		if(res == INF)
			cout<<-1<<"\n";
		else
			cout<<res<<"\n";
	}

	return 0;
}